Add a take_multi! macro to allow the use of multiple variables in a take closure

Ameliorate opened the pull request:
Currently, the macro works up to 5 variables, but can be expanded if need be.

I expected less code to be necessary, but macros are hard, complex, and inexpressive.
# Example

``` rust
struct Foo;
struct Bar;
let mut foo = Foo;
let mut bar = Bar;
take_multi!(&mut foo, &mut bar, |f, b| {    // Brackets must be used.
    // Can access both f and b, later providing a new value.
    drop(f);
    drop(b);
    // Do more stuff
     f = Foo;  // Values are taken out of the closure automatically.
     b = Bar;  // Just set them to the new values.
});
```

Ameliorate committed:
Show what I was trying to do.

Commit:
Add take_multi! macro

Curently, it goes up to 5 arguments. This can be expanded if needed.

Commit:
Merge branch 'master' of github.com:Ameliorate/take_mut

Commit:
Revert "Show what I was trying to do."

This reverts commit 23c0dc2c117ae2d690b9268a9742ef16ced87813.

Sgeo commented:
I was envisioning a non-macro based approach to this problem, but had some trouble with the lifetimes involved.

In the meantime, this looks interesting, but I think I want to play with it a bit. I can't help wonder if there's a way to get any number of `&mut` to work smoothly, perhaps using nested tuples.

Also wondering if the syntax of the macro should use `mut` in the "argument list" to make clear to callers that the binding is mutable.

Ameliorate commented:
`mut` in the arguments list does sound like a good idea for self-documentation. I'll try that when I get back to my computer.

Commit:
Requires that the closure arguments in take_multi! have a mut modifier

The idea is, that since the arguments are made mutable by the macro, the mut self-documents that part of the macro.

Commit:
Makes the body in the take_mut! macro a block instead of a {;*;}

Sgeo commented:
Merged code into take_multi branch instead of using PR.
